File: recipes/jvm/java-recipes/micronaut-postgres-recipe/src/main/java/com/trendyol/stove/recipes/micronaut/application/ProductService.java
package com.trendyol.stove.recipes.micronaut.application;

import com.trendyol.stove.recipes.micronaut.domain.Product;
import com.trendyol.stove.recipes.micronaut.domain.ProductRepository;
import com.trendyol.stove.recipes.micronaut.domain.SupplierService;
import jakarta.inject.Singleton;
import java.util.UUID;
import reactor.core.publisher.Mono;

@Singleton
public class ProductService {

  private final ProductRepository productRepository;
  private final SupplierService supplierService;

  public ProductService(ProductRepository productRepository, SupplierService supplierService) {
    this.productRepository = productRepository;
    this.supplierService = supplierService;
  }

  public Mono<Product> createProduct(String id, String productName, long supplierId) {
    String productId = id != null ? id : UUID.randomUUID().toString();
    return supplierService
        .getSupplierPermission(supplierId)
        .map(supplier ->
            Product.create(productId, productName, supplierId, supplier.isBlacklisted()))
        .flatMap(productRepository::save);
  }
}

File: examples/micronaut-example/src/main/kotlin/stove/micronaut/example/application/services/ProductService.kt
package stove.micronaut.example.application.services

import jakarta.inject.Singleton
import stove.micronaut.example.application.domain.Product
import stove.micronaut.example.application.repository.ProductRepository
import stove.micronaut.example.infrastructure.http.SupplierHttpService

@Singleton
class ProductService(
  private val productRepository: ProductRepository,
  private val supplierHttpService: SupplierHttpService
) {
  suspend fun createProduct(id: String, productName: String, supplierId: Long): Product {
    val supplier = supplierHttpService.getSupplierPermission(supplierId)
    val product = Product.new(id, productName, supplierId, supplier!!.isBlacklisted)
    productRepository.save(product)
    return product
  }
}
